src/raftstore/store/peer.rs

```diff
@@ -1000,77 +996,67 @@ impl Peer {
         cmd_resp::bind_term(&mut resp, self.term());
         cb.call_box((resp,));
 
-        Ok(exec_result)
+        exec_result
     }
 
     pub fn term(&self) -> u64 {
         self.raft_group.raft.term
     }
 
+    // apply operation can fail as following situation:
+    //   1. encouter an error that will occur on all store, it can continue
+    // applying next entry safely;
+    //   2. encouter an error that any not occur on all store, in this case
+    // we should try to apply the entry again or panic. Considering that this
+    // usually due to disk operation fail, which is rare, so just panic is ok.
     fn apply_raft_cmd(&mut self,
                       index: u64,
                       term: u64,
                       req: &RaftCmdRequest)
-                      -> Result<(RaftCmdResponse, Option<ExecResult>)> {
-        if self.pending_remove {
-            let region_not_found = Error::RegionNotFound(self.region_id);
-            let mut resp = cmd_resp::new_error(region_not_found);
-            if let Some(uuid) = util::get_uuid_from_req(req) {
-                cmd_resp::bind_uuid(&mut resp, uuid);
-            }
-            return Ok((resp, None));
-        }
+                      -> (RaftCmdResponse, Option<ExecResult>) {
+        // if pending remove, apply should be aborted already.
+        assert!(!self.pending_remove);
 
         let last_applied_index = self.get_store().applied_index();
         if last_applied_index >= index {
-            return Err(box_err!("applied index moved backwards, {} >= {}",
-                                last_applied_index,
-                                index));
+            // How can it happen?
+            panic!("{} applied index moved backwards, {} >= {}",
+                   self.tag,
+                   last_applied_index,
+                   index);
         }
 
-        let engine = self.engine.clone();
-        let mut ctx = ExecContext {
-            snap: Snapshot::new(engine),
-            apply_state: self.get_store().apply_state.clone(),
-            wb: WriteBatch::new(),
-            req: req,
-            index: index,
-            term: term,
-        };
-        let (mut resp, exec_result) = self.exec_raft_cmd(&mut ctx).unwrap_or_else(|e| {
+        let mut ctx = ExecContext::new(self, index, term, req);
+        let (resp, exec_result) = self.exec_raft_cmd(&mut ctx).unwrap_or_else(|e| {
             error!("{} execute raft command err: {:?}", self.tag, e);
             (cmd_resp::new_error(e), None)
         });
 
         ctx.apply_state.set_applied_index(index);
-        ctx.save(self.region_id).expect("save state must not fail");
+        if !self.pending_remove {
+            assert_op!(self.tag, ctx.save(self.region_id));
```

Comment:
I prefer using unwrap_or_else instead of macro, and panic with a more clear message.